scrapy/utils/console.py

```diff
@@ -1,37 +1,76 @@
     try:
-        try: # use IPython if available
-            if noipython:
-                raise ImportError()
+        from IPython.terminal.embed import InteractiveShellEmbed
+        from IPython.terminal.ipapp import load_default_config
+    except ImportError:
+        from IPython.frontend.terminal.embed import InteractiveShellEmbed
+        from IPython.frontend.terminal.ipapp import load_default_config
+
+    @wraps(_embed_ipython_shell)
+    def wrapper(namespace=namespace, banner=''):
+        config = load_default_config()
+        shell = InteractiveShellEmbed(
+            banner1=banner, user_ns=namespace, config=config)
+        shell()
+    return wrapper
+
+def _embed_bpython_shell(namespace={}, banner=''):
+    """Start a bpython shell"""
+    import bpython
+    @wraps(_embed_bpython_shell)
+    def wrapper(namespace=namespace, banner=''):
+        bpython.embed(locals_=namespace, banner=banner)
+    return wrapper
+
+def _embed_standard_shell(namespace={}, banner=''):
+    """Start a standard python shell"""
+    import code
+    try: # readline module is only available on unix systems
+        import readline
+    except ImportError:
+        pass
+    else:
+        import rlcompleter
+        readline.parse_and_bind("tab:complete")
+    @wraps(_embed_standard_shell)
+    def wrapper(namespace=namespace, banner=''):
+        code.interact(banner=banner, local=namespace)
+    return wrapper
 
+def get_shell_embed_func(shells):
+    """Return the first acceptable shell-embed function
+    from a given list of shell names.
+    """
+    known_shells = {
+        'python' : _embed_standard_shell,
+        'ipython': _embed_ipython_shell,
+        'bpython': _embed_bpython_shell,
+    }
+    for shell in shells:
+        if shell in known_shells and callable(known_shells[shell]):
             try:
-                from IPython.terminal.embed import InteractiveShellEmbed
-                from IPython.terminal.ipapp import load_default_config
+                # function test: run all setup code (imports),
+                # but dont fall into the shell
+                return known_shells[shell]()
             except ImportError:
-                from IPython.frontend.terminal.embed import InteractiveShellEmbed
-                from IPython.frontend.terminal.ipapp import load_default_config
+                continue
 
-            config = load_default_config()
-            shell = InteractiveShellEmbed(
-                banner1=banner, user_ns=namespace, config=config)
+def start_python_console(namespace=None, banner='', shells=None):
+    """Start Python console bound to the given namespace.
+    Readline support and tab completion will be used on Unix, if available.
+    """
+    if namespace is None:
+        namespace = {}
+    if shells is None:
+        shells = ['ipython', 'bpython', 'python']
+
+    try:
+        shell = get_shell_embed_func(shells)
+        if shell is not None:
             shell()
```

Comment:
namespace and banner missing. Before it was being use directly `code.interact(banner=banner, local=namespace)` we should pass it to the shell.